src/core/execute.c

```diff
@@ -419,33 +473,107 @@ int exec_spawn(Unit *unit,
+        rewind(f);
 
-        r = fd_cloexec(fileno(f), false);
-        if (r < 0)
-                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialization fd: %m");
+        if (fds_array) {
+                r = send_many_fds(unit->manager->executors_pool_socket[0], fds_array, n_fds_array, 0);
+                if (r < 0)
+                        return log_unit_error_errno(unit, r, "Failed to send serialization fd: %m");
 
-        r = fdset_cloexec(fdset, false);
-        if (r < 0)
-                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialized fds: %m");
+                r = fd_wait_for_event(unit->manager->executors_pool_socket[0], POLLIN, 1 * USEC_PER_SEC);
+                if (r <= 0) {
+                        log_unit_debug(unit, "No response from any worker, clearing out the pool.");
 
-        r = log_level_to_string_alloc(log_get_max_level(), &log_level);
-        if (r < 0)
-                return log_unit_error_errno(unit, r, "Failed to convert log level to string: %m");
+                        /* Something's very wrong, there are workers in the pool, but there is no response?
+                         * Kill the slackers (to ensure no double processing) and then fall back to a direct
+                         * spawn. The manager will respawn workers later. */
+                        (void) manager_kill_workers(unit->manager);
 
-        xsprintf(serialization_fd_number, "%i", fileno(f));
+                        /* Let's REALLY make sure. Now the cgroup is empty, so if a worker managed to sneak
+                         * past somehow between the poll and the kill, we'll have a message in the pipe. */
+                        r = fd_wait_for_event(unit->manager->executors_pool_socket[0], POLLIN, 0);
+                }
+                if (r <= 0) {
+                        log_unit_debug(unit, "Falling back to direct spawn.");
+
+                        /* Now that we know there are no messages in the socket and all workers are down,
+                         * also close the sockets so that any message we queued before is lost, otherwise
+                         * the next time workers are started, the message queued now that is taken over by
+                         * this direct execution would be picked up. */
+                        safe_close_pair(unit->manager->executors_pool_socket);
+                        if (socketpair(AF_UNIX,
+                                       SOCK_DGRAM|SOCK_CLOEXEC,
+                                       /* protocol= */ 0,
+                                       unit->manager->executors_pool_socket) < 0)
+                                return log_unit_error_errno(unit, errno, "Failed to open socket pair: %m");
+
+                        /* Unfortunately this means we have to redo the serialization, as the FDs will now
+                         * be passed through directly, rather than via a socket message. */
```

Comment:
why though? why not spawn the child with a socketpair() socket as fd3 or so, with the relevant datagram queued into it?

Reply:
i'd probably not bother with the fallback btw. and instead just fail completely...

Reply:
by passing a socket with a relevant datagram in it to the child you have the benefit that you can send a single fd to the child, and one day when glibc supports it tell it to close all others. in the child. that's much nicer than having to give it a scattered list of fds to keep open

Reply:
Something might have consumed it, and exploded before returning, so there's no guarantee the message will still be there. It's a fallback path, so I think it's fine to re-do it. Also it's all new, and this handover is new and more fragile than the previous approach, so I think having a fallback is important in case things go wrong with the handover.